k-distribution/src/main/scripts/lib/pyk/pyk/cterm.py

```diff
@@ -0,0 +1,43 @@
+from dataclasses import dataclass
+from functools import cached_property
+from itertools import chain
+from typing import List, Optional, Tuple
+
+from .kast import KInner, flattenLabel
+from .kastManip import match, splitConfigAndConstraints
+from .prelude import mlAnd
+from .subst import Subst
+
+
+@dataclass(frozen=True)
+class CTerm:
+    config: KInner  # TODO Optional?
+    constraints: Tuple[KInner, ...]
+
+    def __init__(self, cterm: KInner) -> None:
+        config, constraint = splitConfigAndConstraints(cterm)
+        constraints = tuple(flattenLabel('#And', constraint))
+        object.__setattr__(self, 'config', config)
+        object.__setattr__(self, 'constraints', constraints)
+
+    def __iter__(self):
+        return chain([self.config], self.constraints)
+
+    @cached_property
+    def cterm(self) -> KInner:
+        return mlAnd(self)
+
+    @property
+    def hash(self) -> str:
+        return self.cterm.hash
+
+    def match(self, pattern: 'CTerm') -> Optional[Tuple[Subst, List[KInner]]]:
+        subst = match(pattern=pattern.config, term=self.config)
+
+        if subst is None:
+            return None
+
+        assumptions = set(self.constraints)
+        obligations = [constraint for constraint in map(subst, pattern.constraints) if constraint not in assumptions]
+
+        return subst, obligations
```

Comment:
The obligations here need to be `#Top`, correct? Can we document this?

Reply:
This method can be considered a generalization of `kastManip.matchWithConstraint` (though it's not as smart yet, so we can improve that later). If `obligations` is empty, it indicates a syntactic match. Otherwise, it contains the obligations that need to be shown (assuming `self.constraints`) for the implication to hold.

Reply:
Well, specifically the obligations need to be shown _valid_, not just _satisfiable_. That is, `simplify(obligations)` must be `#Top`, it's not enough that this is a model of the remaining obligations, there must be no models of the negations of the obligatios (`#Not obligations` must be `#Bottom`, unsatisfiable).

Otherwise we are doing some sort of hybrid between matching on the configuration and unification on the constraints. It seems dangerous to expose this method as it is to me, because it is some weird hybrid of matching and unification. We should eventually support some limited unification, but in that case we should be doing unification of both the configuration and the constraints.

I would rather fail with a fatal error if `len(obligations) > 0` for the moment, because (i) thats good enough for the hacky version of the tool as it is right now because we make sure to maintain syntactic constraint stability externally anyway, and (ii) we can relax this restriction later as we add more semantic information (better to start with limited safe version and expand functionality).

Reply:
So, for example, if `3 >Int X` was in `obligations`, then it is not valid because `X ==Int 5` doesn't work, but it is satisfiable because `X ==Int 2` does work. This means matching should not succeed.

Reply:
> Well, specifically the obligations need to be shown valid

Definitely. What the `kit` tool does for example is construct a claim for the validity of `(/\ self.constraints) -> (/\ obligations)`. This is necessary as otherwise even the simplest nontrivial example (sum-to-n) fails to converge.

> It seems dangerous to expose this method as it is to me, because it is some weird hybrid of matching and unification.

How about I add a docstring elaborating on how `obligations` is supposed to be interpreted? Then each caller can use it according to its needs, i.e. fail if `len(obligations) > 0`, or do something with it otherwise. I think it is a useful function is this context.

Alternatively, I can add both the syntactic and semantic versions of `match`, with the syntactic one being a special case. I can also remove it, but then I'll have tor reproduce it in `kit` (and will probably end up upstreaming it eventually).

Reply:
Changed to return a matching logic formula instead of a list.